src/librustdoc/html/highlight.rs

```diff
@@ -440,12 +540,91 @@ fn exit_span(out: &mut Buffer) {
 ///     enter_span(Foo), string("text", None), exit_span()
 ///     string("text", Foo)
 /// ```
+///
 /// The latter can be thought of as a shorthand for the former, which is more
 /// flexible.
-fn string<T: Display>(out: &mut Buffer, text: T, klass: Option<Class>) {
+///
+/// Note that if `context` is not `None` and that the given `klass` contains a `Span`, the function
+/// will then try to find this `span` in the `span_correspondance_map`. If found, it'll then
+/// generate a link for this element (which corresponds to where its definition is located).
+fn string<T: Display>(
+    out: &mut Buffer,
+    text: T,
+    klass: Option<Class>,
+    context: Option<&Context<'_>>,
+    root_path: &str,
+) {
     match klass {
         None => write!(out, "{}", text),
-        Some(klass) => write!(out, "<span class=\"{}\">{}</span>", klass.as_html(), text),
+        Some(klass) => {
+            if let Some(def_span) = klass.get_span() {
+                let mut text = text.to_string();
+                if text.contains("::") {
+                    text =
+                        text.split("::").enumerate().fold(String::new(), |mut path, (pos, t)| {
+                            let pre = if pos != 0 { "::" } else { "" };
+                            match t {
+                                "self" | "Self" => write!(
+                                    &mut path,
+                                    "{}<span class=\"{}\">{}</span>",
+                                    pre,
+                                    Class::Self_((0, 0)).as_html(),
+                                    t
+                                ),
+                                "crate" | "super" => write!(
+                                    &mut path,
+                                    "{}<span class=\"{}\">{}</span>",
+                                    pre,
+                                    Class::KeyWord.as_html(),
+                                    t
+                                ),
+                                t => write!(&mut path, "{}{}", pre, t),
+                            }
+                            .expect("Failed to build source HTML path");
+                            path
+                        });
+                }
+                if let Some(context) = context {
+                    if let Some(href) =
+                        context.shared.span_correspondance_map.get(&def_span).and_then(|href| {
+                            match href {
+                                LinkFromSrc::Local(span) => {
+                                    // FIXME: this is a VERY weird bug where `SpanInterner` (which
+                                    // is used inside `Span`) receives an invalid index.
+                                    if !span.is_valid_data() {
```

Comment:
Can you give an example of when this happens? Maybe a backtrace? I don't think the compiler should *ever* have invalid spans, that seems like a bug.

Reply:
This is definitely a bug because the spans are valid when I put them into the `span_map` (tested and no problem). This bug is triggered when using this feature on libcore. As for where it comes from: https://github.com/rust-lang/rust/blob/c20c9219dae5ea56ce9bf1c211fafdc7da8700b9/compiler/rustc_span/src/span_encoding.rs#L123

As you can see, the index created to access the span is out-of-bound (for reasons I couldn't determine) whereas it works perfectly well when I'm building the `span_map`. cc @estebank

Reply:
This is fixed now, right?

Reply:
Yes it is, which is why it doesn't appear in the diff anymore. ;)